drivers/storage/openstack/storage/openstack_storage.go

```diff
@@ -0,0 +1,743 @@
+	}
+
+	return libstorageSnapshots, nil
+}
+
+func translateSnapshot(snapshot *snapshots.Snapshot) *types.Snapshot {
+	return &types.Snapshot{
+		Name:        snapshot.Name,
+		VolumeID:    snapshot.VolumeID,
+		ID:          snapshot.ID,
+		VolumeSize:  int64(snapshot.Size),
+		StartTime:   time.Time(snapshot.CreatedAt).Unix(),
+		Description: snapshot.Description,
+		Status:      snapshot.Status,
+	}
+}
+
+func (d *driver) VolumeSnapshot(
+	ctx types.Context,
+	volumeID, snapshotName string,
+	opts types.Store) (*types.Snapshot, error) {
+
+	fields := eff(map[string]interface{}{
+		"snapshotName": snapshotName,
+		"volumeId":     volumeID,
+	})
+
+	createOpts := snapshots.CreateOpts{
+		Name:     snapshotName,
+		VolumeID: volumeID,
+		Force:    true,
+	}
+
+	snapshot, err := snapshots.Create(d.clientBlockStorage, createOpts).Extract()
+	if err != nil {
+		return nil,
+			goof.WithFieldsE(fields, "error creating snapshot", err)
+	}
+
+	ctx.WithFields(fields).Info("waiting for snapshot creation to complete")
+
+	err = snapshots.WaitForStatus(d.clientBlockStorage, snapshot.ID, "available", int(d.volumeSnapshotTimeout().Seconds()))
+	if err != nil {
+		return nil,
+			goof.WithFieldsE(fields,
+				"error waiting for snapshot creation to complete", err)
+	}
+
+	return translateSnapshot(snapshot), nil
+
+}
+
+func (d *driver) SnapshotRemove(
+	ctx types.Context,
+	snapshotID string,
+	opts types.Store) error {
+	resp := snapshots.Delete(d.clientBlockStorage, snapshotID)
+	if resp.Err != nil {
+		return goof.WithFieldsE(goof.Fields{
+			"snapshotId": snapshotID}, "error removing snapshot", resp.Err)
+	}
+
+	return nil
+}
+
+func (d *driver) VolumeCreate(ctx types.Context, volumeName string,
+	opts *types.VolumeCreateOpts) (*types.Volume, error) {
+
+	return d.createVolume(ctx, volumeName, "", "", opts)
+}
+
+func (d *driver) VolumeCreateFromSnapshot(
+	ctx types.Context,
+	snapshotID, volumeName string,
+	opts *types.VolumeCreateOpts) (*types.Volume, error) {
+
+	return d.createVolume(ctx, volumeName, "", snapshotID, opts)
+}
+
+func (d *driver) VolumeCopy(
+	ctx types.Context,
+	volumeID, volumeName string,
+	opts types.Store) (*types.Volume, error) {
+	volume, err := d.VolumeInspect(ctx, volumeID, &types.VolumeInspectOpts{})
+	if err != nil {
+		return nil,
+			goof.New("error getting reference volume for volume copy")
+	}
+
+	volumeCreateOpts := &types.VolumeCreateOpts{
+		Type:             &volume.Type,
+		AvailabilityZone: &volume.AvailabilityZone,
+	}
+
+	return d.createVolume(ctx, volumeName, volumeID, "", volumeCreateOpts)
+}
+
+func (d *driver) createVolume(
+	ctx types.Context,
+	volumeName string,
+	volumeSourceID string,
+	snapshotID string,
+	opts *types.VolumeCreateOpts) (*types.Volume, error) {
+
+	options := &volumes.CreateOpts{
+		Name:             volumeName,
+		SnapshotID:       snapshotID,
+		SourceReplica:    volumeSourceID,
+	}
+
+	if opts.Type != nil {
+		options.VolumeType = *opts.Type
+	}
+	if opts.Size != nil {
```

Comment:
Please protect against the creation of size zero volumes in the style of https://github.com/codedellemc/libstorage/pull/495.  I know it's a pain that mass driver changes have happened while you've been working on this.